---
title: "Intro to Epigenomics"
author: Wade Boohar
date: 11/03/24
updated: 03/07/24
User: Alex Zhang
---


```{r setup}
 knitr::opts_knit$set(root.dir = normalizePath("/home1/alexz/490_cluster/analysis_data"))
```

Package Download and Data-cleaning
```{r}
if (!require("sesameData", quietly = TRUE))
BiocManager::install("sesameData")

if (!require("sesame", quietly = TRUE))
BiocManager::install("sesame")

if (!require("limma", quietly = TRUE))
BiocManager::install("limma")
```


Load in all necessary packages
```{r}
library(TCGAbiolinks)
library(sesame)
library(sesameData)
library(limma)
```
```{r}
knitr::opts_knit$set(root.dir = normalizePath("/project/rohs_1070/analysis_data"))
```

```{r}
methylation_clinical <- read.csv("/project/rohs_1070/analysis_data/brca_methylation_clinical.csv")
methylation_betas <- read.csv("/project/rohs_1070/analysis_data/brca_methylation_betas.csv")

methylation_cpg <- read.csv("/project/rohs_1070/analysis_data/brca_cpg_sites.csv")


# query <- GDCquery(project = "TCGA-BRCA",
#                   data.category = "DNA Methylation",
#                   data.type = "Methylation Beta Value",
#                   platform = "Illumina Human Methylation 450",
#                   data.format = "Dataframe")
# #lGDCdownload(query)
# methylation450 <- GDCprepare(query)
```

```{r}
# methylation_clinical <- as.data.frame(methylation450@colData)
# betas <- as.data.frame(methylation450@assays@data@listData)
# cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)

column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type','sites_of_involvement'), F, T)

methylation_clinical <- methylation_clinical[,column_mask]
write.csv(methylation_clinical, 'brca_methylation_clinical.csv')


site_mask <- !grepl('-', methylation_cpg$gene) & !grepl(';', methylation_cpg$gene) & !is.na(methylation_cpg$gene) & complete.cases(methylation_betas)
methylation_betas <- methylation_betas[site_mask,]
methylation_cpg <- methylation_cpg[site_mask,]

knitr::opts_knit$set(root.dir = normalizePath("/home1/alexz/490_cluster/week8_Methylation"))
write.csv(methylation_betas, 'brca_methylation_betas.csv')
write.csv(methylation_cpg, 'brca_cpg_sites.csv')
```

```{r}
methylation_clinical <- read.csv('brca_methylation_clinical.csv', row.names = 1)
betas <- read.csv('brca_methylation_betas.csv', row.names = 1)
cpg_sites <- read.csv('brca_cpg_sites.csv', row.names = 1)
```


(1) Naive Differential Methylation
```{r}
#masking out NAs
na_mask <- !is.na(methylation_clinical$definition)
methylation_clinical <- methylation_clinical[na_mask,]
betas_clean <- betas[,na_mask]
beta_clean <- betas_clean[,-1]

methylation_clinical$definition <- ifelse(methylation_clinical$definition == "Metastatic", "Metastatic", "Non_Metastatic")

#fitting linear models using a "target value"
metastatic_mask <- methylation_clinical$definition == "Metastatic"

methylation_clinical$metastatic <- !metastatic_mask

mval <- t(apply(beta_clean, 1, function(x) log2(x/(1-x)))) 
#mvalue is another statistic for methylation, centered at 0 and ranges from -1 to 1
design <- model.matrix(~metastatic, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
# dim(mval)           # Should be: features (rows) x samples (columns)
# dim(design)         # Should be: samples (rows) x model variables (columns)
# # Subset and reorder to match mval sample order
# methylation_clinical <- methylation_clinical[colnames(mval), ]
```


```{r}
#Extracting model into dataframe
dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
```


(2) Direct comparison of methylation status to transcriptional activity

#...INSERT DESeq2 Stuff here to generate 'results'...
```{r}
knitr::opts_knit$set(root.dir = normalizePath("/home1/alexz/490_cluster/analysis_data"))
```

```{r}
library(TCGAbiolinks)
library(DESeq2)
library(ggplot2)

```

```{r}
rna_query <- GDCquery(project ="TCGA-SKCM",
data.category = "Transcriptome Profiling",
data.type = "Gene Expression Quantification",
workflow.type = "STAR - Counts")
#GDCdownload(rna_query)
rna_se <- GDCprepare(rna_query)
```

```{r}
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)
rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)
rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)
```

```{r}
#colnames(rna_clinical)
rna_clinical <- rna_clinical[,-23]
unique(rna_clinical$definition)

```

```{r}
row.names(rna_genes) <- rna_genes$gene_id
row.names(rna_counts) <- rna_genes$gene_id #give the rna count data row and col names
colnames(rna_counts) <- rna_clinical$barcode

# Filter out samples with missing definition (Metastatic status)
metastatic_mask <- !is.na(rna_clinical$definition)
clean_clinical <- rna_clinical[metastatic_mask, ]

clean_clinical$definition <- ifelse(clean_clinical$definition == "Metastatic","Metastatic", "Non_Metastatic")

# Convert to factor (optional but recommended for modeling)
clean_clinical$definition <- factor(clean_clinical$definition)
clean_clinical$definition <- relevel(clean_clinical$definition, ref = "Metastatic")

# Filter out samples with missing age
age_mask <- !is.na(clean_clinical$age_at_index)
clean_clinical <- clean_clinical[age_mask, ]

# Create age category column
clean_clinical$age_category <- ifelse(clean_clinical$age_at_index >= 58, "old", "young")
clean_clinical$age_category <- factor(clean_clinical$age_category)

# Match RNA counts with the filtered clinical data
clean_counts <- rna_counts[, metastatic_mask]
clean_counts <- clean_counts[, age_mask]

# Filter out genes with total counts less than 20
filter_mask <- rowSums(clean_counts) >= 20
clean_counts <- clean_counts[filter_mask, ]
clean_genes <- rna_genes[filter_mask, ]

# Factor the gender column
clean_clinical$gender <- factor(clean_clinical$gender)

dds <- DESeqDataSetFromMatrix(countData = clean_counts,
                              colData = clean_clinical,
                              design=~ age_category + gender + definition)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, contrast = c("definition", "Non_Metastatic", "Metastatic"))

results <- data.frame(results)
```

```{r}
#you can also try looking at "upregulated" or "hypermethylated" !
downregulated <- results[(results$log2FoldChange < 3), 'gene_name']
hypomethylated <- dat[dat$foldchange < -1, 'geneName']
interest_genes <- intersect(downregulated, hypomethylated)
```


(Extra) Making Boxplots
```{r}
GENE <- "MAD1L1"

# Create masks for the gene of interest
gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

# Create masks for metastatic vs non-metastatic
rna_clinical_metastatic <- rna_clinical$definition == "Metastatic"
rna_clinical_nonmetastatic <- rna_clinical$definition == "Non_Metastatic"

# Apply same filtering to methylation data
methylation_clinical_metastatic <- methylation_clinical$definition == "Metastatic"
methylation_clinical_nonmetastatic <- methylation_clinical$definition == "Non_Metastatic"

# Extract RNA counts and methylation betas
rna_metastatic <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_metastatic])
rna_nonmetastatic <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_nonmetastatic])

methylation_metastatic <- betas[gene_betas_mask, methylation_clinical_metastatic]
methylation_nonmetastatic <- betas[gene_betas_mask, methylation_clinical_nonmetastatic]

# Boxplot for RNA expression
boxplot(rna_nonmetastatic, rna_metastatic,
        xlab = 'Group', ylab = 'Counts',
        names = c('Non_Metastatic', 'Metastatic'))

```
3.After looking at the UCSC Genome Browser, I observed that for genes such as COL23A1, MASP2, and LTBP4, several CpG sites are located very close to their promoter regions. This positioning suggests potential epigenetic regulation through DNA methylation, which could influence their transcriptional activity.â€‹

Specifically, studies have demonstrated that the LTBP4 gene exhibits differential methylation patterns within its promoter-associated CpG islands, leading to variations in gene expression levels. For instance, certain CpG islands within the LTBP4 promoter show varying degrees of methylation, which correlates with altered transcriptional activity in different cell lines. (Tan, Ming, et al. "Characterization of LTBP4 variants and their role in pulmonary disease." PLOS ONE, vol. 8, no. 6, 2013, e65614. https://doi.org/10.1371/journal.pone.0065614.)  

